fpga_interchange/dedicated_interconnect.cc

```diff
@@ -0,0 +1,785 @@
+enum WireNodeState
+{
+    IN_SINK_SITE = 0,
+    IN_ROUTING = 1,
+    IN_SOURCE_SITE = 2
+};
+
+struct WireNode
+{
+    WireId wire;
+    WireNodeState state;
+    int depth;
+};
+
+// Maximum depth that a dedicate interconnect is considered.
+//
+// Routing networks with depth <= kMaxDepth is considers a dedicated
+// interconnect.
+constexpr int kMaxDepth = 20;
+
+void DedicatedInterconnect::init(const Context *ctx)
+{
+    this->ctx = ctx;
+
+    if (ctx->debug) {
+        log_info("Finding dedicated interconnect!\n");
+    }
+
+    find_dedicated_interconnect();
+    if (ctx->debug) {
+        print_dedicated_interconnect();
+    }
+}
+
+bool DedicatedInterconnect::check_routing(BelId src_bel, IdString src_bel_pin, BelId dst_bel,
+                                          IdString dst_bel_pin) const
+{
+    std::vector<WireNode> nodes_to_expand;
+
+    WireId src_wire = ctx->getBelPinWire(src_bel, src_bel_pin);
+
+    const auto &src_wire_data = ctx->wire_info(src_wire);
+    NPNR_ASSERT(src_wire_data.site != -1);
+
+    WireId dst_wire = ctx->getBelPinWire(dst_bel, dst_bel_pin);
+
+    const auto &dst_wire_data = ctx->wire_info(dst_wire);
+    NPNR_ASSERT(dst_wire_data.site != -1);
+
+    WireNode wire_node;
+    wire_node.wire = src_wire;
+    wire_node.state = IN_SOURCE_SITE;
+    wire_node.depth = 0;
+
+    nodes_to_expand.push_back(wire_node);
+
+    while (!nodes_to_expand.empty()) {
+        WireNode node_to_expand = nodes_to_expand.back();
+        nodes_to_expand.pop_back();
+
+        for (PipId pip : ctx->getPipsDownhill(node_to_expand.wire)) {
+            if (ctx->is_pip_synthetic(pip)) {
+                continue;
+            }
+
+            WireId wire = ctx->getPipDstWire(pip);
+            if (wire == WireId()) {
+                continue;
+            }
+
+#ifdef DEBUG_EXPANSION
+            log_info(" - At wire %s via %s\n", ctx->nameOfWire(wire), ctx->nameOfPip(pip));
+#endif
+
+            WireNode next_node;
+            next_node.wire = wire;
+            next_node.depth = node_to_expand.depth += 1;
+
+            if (next_node.depth > kMaxDepth) {
+                // Dedicated routing should reach sources by kMaxDepth (with
+                // tuning).
+                //
+                // FIXME: Consider removing kMaxDepth and use kMaxSources?
+                return false;
+            }
+
+            auto const &wire_data = ctx->wire_info(wire);
+
+            bool expand_node = true;
+            if (ctx->is_site_port(pip)) {
+                switch (node_to_expand.state) {
+                case IN_SOURCE_SITE:
+                    NPNR_ASSERT(wire_data.site == -1);
+                    next_node.state = IN_ROUTING;
+                    break;
+                case IN_ROUTING:
+                    NPNR_ASSERT(wire_data.site != -1);
+                    if (wire.tile == src_wire.tile && wire_data.site == src_wire_data.site) {
+                        // Dedicated routing won't have straight loops,
```

Comment:
Are we sure this is a universal truth? It does seem like a good heuristic, but I'm wondering if there are places where combinations of routethroughs/loopbacks might violate this.

Reply:
I'd need to see an example of where a straight loop forms a dedicated interconnect (e.g. placement dependent).  All the cases I know are site to another site.  Do you know of any counter examples?

The good news is if this logic fails on an arch, it will manifest eventually as a routing failure, which will generate a bug report, rather than an optimization failure.